tensorflow-frameworks/tensorflow-data/src/test/java/org/tensorflow/data/DatasetOpTester.java

```diff
@@ -0,0 +1,283 @@
+            Shape.of(BATCH_SIZE, 1)));
+
+        MakeIterator makeIterator = tf.data.makeIterator(batchDataset, anonymousIter);
+
+        session.runner()
+            .addTarget(makeIterator.op())
+            .run();
+
+        IteratorGetNext getNext = tf.data.iteratorGetNext(anonymousIter, outputTypes, outputShapes);
+        Operand<?> XOp = getNext.components().get(0);
+        Operand<?> yOp = getNext.components().get(1);
+        while (true) {
+          try {
+            List<Tensor<?>> outputs = session.runner()
+                .addTarget(getNext.op())
+                .fetch(XOp)
+                .fetch(yOp)
+                .run();
+            System.out.println("BATCH: ");
+            printIntTensor(outputs.get(0));
+            printIntTensor(outputs.get(1));
+            System.out.println();
+          } catch (IndexOutOfBoundsException e) {
+            System.out.println("finished iterating.");
+            break;
+          }
+        }
+      }
+    }
+
+  }
+
+  @Test
+  public void testCleanGraphBatching() {
+    try (Graph graph = new Graph()) {
+
+      Ops tf = Ops.create(graph);
+      try (Session session = new Session(graph)) {
+        long BATCH_SIZE = 2L;
+
+        Constant<TInt32> X = tf.val(
+            new int[][]{
+                {1, 2, 3},
+                {4, 5, 6},
+                {7, 8, 9},
+                {10, 11, 12}
+            }
+        );
+
+        Constant<TInt32> y = tf.val(
+            new int[][]{
+                {1},
+                {4},
+                {7},
+                {10}
+            }
+        );
+
+        // // Try running TensorDataset
+        List<Operand<?>> tensors = Arrays.asList(X, y);
+        List<DataType<?>> outputTypes = Arrays.asList(TInt32.DTYPE, TInt32.DTYPE);
+
+        Dataset dataset = Dataset
+            .fromTensorSlices(tf, tensors, outputTypes)
+            .batch(2);
+
+        OneShotIterator oneShotIterator = dataset.makeOneShotIterator();
+        List<Output<?>> components = oneShotIterator.getComponents();
+        Operand<?> XOp = components.get(0);
+        Operand<?> yOp = components.get(1);
+
+        // Run MakeIterator Op
+        session.runner()
+            .addTarget(oneShotIterator.getMakeIteratorOp())
+            .run();
+
+        while (true) {
+          try {
+            List<Tensor<?>> outputs = session.runner()
+                .fetch(XOp)
+                .fetch(yOp)
+                .run();
+
+            System.out.println("BATCH: ");
+            printIntTensor(outputs.get(0));
+            printIntTensor(outputs.get(1));
+            System.out.println();
+          } catch (IndexOutOfBoundsException e) {
+            System.out.println("finished iterating.");
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  public static void printIntTensor(Tensor<?> tensor) {
+    IntBuffer buffer = IntBuffer.allocate((int) tensor.shape().size());
+    tensor.writeTo(buffer);
```

Comment:
`tensor.writeTo` might be remove soon, you might want to use `tensor.data()` instead and maybe enhanced `NdArray` with what is missing (like `ndArray.print()`?)

Reply:
I removed the print statements; but it would be great to have an `print` statement for `NdArrays`!